src/app/network/DownloadManager.ts

```diff
@@ -0,0 +1,170 @@
+        if (task?.status !== TaskStatus.Cancelled) {
+          tasksService.updateTask({
+            taskId,
+            merge: {
+              status: TaskStatus.InProcess,
+              nItems: (task?.nItems ?? 0) + 1,
+            },
+          });
+        }
+      };
+
+      tasksService.updateTask({
+        taskId: taskId,
+        merge: {
+          status: TaskStatus.Decrypting,
+        },
+      });
+
+      if (items.length > 1) {
+        await DownloadManagerService.instance.downloadItems(downloadTask, updateProgressCallback, incrementItemCount);
+      } else if (items[0].isFolder) {
+        await DownloadManagerService.instance.downloadFolder(downloadTask, updateProgressCallback, incrementItemCount);
+      } else {
+        await DownloadManagerService.instance.downloadFile(downloadTask, updateProgressCallback);
+      }
+
+      tasksService.updateTask({
+        taskId,
+        merge: {
+          status: TaskStatus.Success,
+        },
+      });
+    } catch (err) {
+      this.reportError(err, downloadTask);
+      throw err;
+    } finally {
+      tasksService.removeListener({ event: TaskEvent.TaskCancelled, listener: cancelTaskListener });
+    }
+  };
+
+  private static readonly reportError = (err: unknown, downloadTask: DownloadTask) => {
+    const { items, taskId } = downloadTask;
+
+    if (err instanceof ConnectionLostError) {
+      tasksService.updateTask({
+        taskId,
+        merge: { status: TaskStatus.Error, subtitle: t('error.connectionLostError') as string },
+      });
+      throw err;
+    }
+
+    const task = tasksService.findTask(taskId);
+
+    if (task !== undefined && task.status !== TaskStatus.Cancelled) {
+      if (items.length > 1) {
+        errorService.reportError(err);
+      } else {
+        const item = items[0];
+        if (item.isFolder) {
+          errorService.reportError(err, {
+            extra: { folder: item.name, bucket: item.bucket, folderParentId: item.parentId },
+          });
+        } else {
+          errorService.reportError(err, {
+            extra: { fileName: item.name, bucket: item.bucket, fileSize: item.size, fileType: item.type },
+          });
+        }
+      }
+      tasksService.updateTask({
+        taskId,
+        merge: {
+          status: TaskStatus.Error,
+        },
+      });
+
+      const castedError = errorService.castError(err);
+
+      if (downloadTask.options.showErrors) {
+        const errorText = items.length > 1 || !items[0].isFolder ? 'error.downloadingFile' : 'error.downloadingFolder';
+
+        notificationsService.show({
+          text: t(errorText, { message: castedError.message || '' }),
+          type: ToastType.Error,
+        });
+      }
+    } else {
+      tasksService.updateTask({
+        taskId,
+        merge: {
+          status: TaskStatus.Cancelled,
+        },
+      });
+    }
+  };
+
+  static readonly downloadItem = async (downloadItem: DownloadItem) => {
```

Comment:
Same comment as the `UploadManager`, it is not necessary to do it now, but it would be good to use a Factory pattern or something that fits to divide all the direct responsibilities that the ‘Managers’ have. 
This way the testing would have been simpler, it would have been divided in several files instead of making one 700 lines file.